client/components/webpack-build-monitor/index.tsx

```diff
@@ -0,0 +1,222 @@
+	const [ lastHash, setLastHash ] = useState( __webpack_hash__ );
+
+	// Webpack
+	const [ isWebpackConnected, setIsWebpackConnected ] = useState( false );
+	const [ hasWebpackErrors, setHasWebpackErrors ] = useState( false );
+	const [ hasWebpackWarnings, setHasWebpackWarnings ] = useState( false );
+	const [ isWebpackBuilding, setWebpackBuilding ] = useState( false );
+
+	// CSS
+	const [ isCssConnected, setIsCssConnected ] = useState( false );
+	const [ hasCssErrors, setHasCssErrors ] = useState( false );
+	const [ isCssBuilding, setCssBuilding ] = useState( false );
+
+	useEffect( () => {
+		const namespace = location.protocol + '//' + location.host + '/css-hot-reload';
+
+		let socket: SocketIOClient.Socket | undefined;
+
+		const connect = () => {
+			debug( 'Hot CSS connecting' );
+			socket = io.connect( namespace, {
+				reconnectionDelay: CONNECTION_TIMEOUT,
+				transports: [ 'websocket' ],
+			} );
+			socket.on( 'connect', () => {
+				debug( 'Hot CSS connected' );
+				setIsCssConnected( true );
+			} );
+
+			socket.on( 'disconnect', () => {
+				debug( 'Hot CSS disconnected. Reconnecting…' );
+				setIsCssConnected( false );
+			} );
+
+			socket.on(
+				'css-hot-reload',
+				( { status }: { status: 'building' | 'build-failed' | 'reload' | 'up-to-date' } ) => {
+					debug( 'Hot CSS status update: %o', status );
+					switch ( status ) {
+						case 'building':
+							setCssBuilding( true );
+							break;
+
+						case 'build-failed':
+							setHasCssErrors( true );
+							setCssBuilding( false );
+							break;
+
+						case 'reload':
+						case 'up-to-date':
+							setHasCssErrors( false );
+							setCssBuilding( false );
+							break;
+					}
+				}
+			);
+		};
+
+		connect();
+
+		return () => {
+			if ( socket && ! socket.disconnected ) {
+				socket.close();
+			}
+		};
+	}, [] );
+
+	useEffect( () => {
+		if ( typeof EventSource === 'undefined' ) {
+			if ( process.env.NODE_ENV !== 'production' ) {
+				// eslint-disable-next-line no-console
+				console.warn( 'Webpack build monitor disabled. No `EventSource`.' );
+			}
+		} else {
+			let lastActivity: TimestampMS;
+			let connectionTimer: IntervalHandle;
+			let source: EventSource;
+			const connect = () => {
+				debug( 'Webpack HMR connecting' );
+				source = new EventSource( '/__webpack_hmr' );
+
+				source.onopen = () => {
+					debug( 'Webpack HMR connected' );
+					setIsWebpackConnected( true );
+					lastActivity = Date.now();
+					connectionTimer = setInterval( () => {
+						if ( Date.now() - lastActivity > CONNECTION_TIMEOUT ) {
+							debug( 'Webpack HMR connection timeout. Reconnecting in %o…', CONNECTION_TIMEOUT );
+							setIsWebpackConnected( false );
+							clearInterval( connectionTimer );
+							source.close();
+							setTimeout( connect, CONNECTION_TIMEOUT );
+						}
+					}, CONNECTION_TIMEOUT / 2 );
+				};
+
+				source.onmessage = ( m: MessageEvent ) => {
+					lastActivity = Date.now();
+					if ( m.data !== '💓' ) {
```

Comment:
I'd love to know the reasons behind this amazing line 😄

Reply:
It's the heartbeat!

https://github.com/webpack-contrib/webpack-hot-middleware/blob/cb29abb9dde435a1ac8e9b19f82d7d36b1093198/middleware.js#L71

Reply:
Hah, amazing! Thank you 🙂